src/rules/no-classfield-shadowing.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * @fileoverview Disallows properties shadowed as class fields
+ * @author Michel Langeveld <https://github.com/michellangeveld>
+ */
+
+import {Rule} from 'eslint';
+import * as ESTree from 'estree';
+import {getClassFields, getPropertyMap} from '../util';
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+const rule: Rule.RuleModule = {
+  meta: {
+    docs: {
+      description: 'Disallows properties shadowed as class field',
+      recommended: true,
+      url: 'https://github.com/43081j/eslint-plugin-lit/blob/master/docs/rules/no-classfield-shadowing.md'
+    },
+    schema: [],
+    messages: {
+      noClassfieldShadowing:
+        'The {{ prop }} property is a class field which has the same as ' +
+        'static property which could have unintended side-effects.'
+    }
+  },
+
+  create(context): Rule.RuleListener {
+    return {
+      ClassDeclaration: (node: ESTree.Class): void => {
+        const propertyMap = getPropertyMap(node);
+        const classMembers = getClassFields(node);
+
+        for (const [prop, {key}] of propertyMap.entries()) {
+          if (classMembers.has(prop)) {
```

Comment:
here or in the util, we also need to account for `declare`. i.e. a property definition in tseslint has `declare: true` in this example:

```ts
class Foo extends LitElement {
  declare foo: string;

  static properties = {foo: {type: String}};
}
```

this is valid since it doesn't result in an actual class field being defined, only declared

to get at the type, you might need to cast as `TSESTree.PropertyDefinition` (since `declare` won't exist on estree's normal interface i imagine). you may need to add `@typescript-eslint/typescript-estree` as a dev dependency.

Reply:
Just back from holiday, will look into the comments.

Reply:
It looks like the parser used by the plugin doesnt support `declare`, im guessing we'd have to move the plugin to use the TS parser, but that seems like a more involved/impactful change. @43081j , Considering the `declare` usecase is a bit of an edgecase, can we do this in a separate PR instead, so that we can unblock michel for the current PR?

Reply:
the problem with that is that the docs recommend using `declare` to solve this problem in TS. so if we ship this rule as-is, it will see that as an error even though the docs recommend it.

i think temporarily you should cast as an `estree.PropertyDefinition & {declare?: boolean}`. leave a TODO stating we should cast as the equivalent tsestree `PropertyDefinition` or something.

Reply:
Its not about TS types, the `declare` keyword is not present on the AST

Reply:
it is. the `declare` keyword is in the AST when the consumer uses teseslint with our plugin

you just don't have it in the tests, which is something to be solved in another PR.

a consumer can still load it up in their eslint and that is the AST we will be given (which i happen to know has `declare?: boolean` in its prop definition).

Reply:
but that then forces consumers of the plugin to also use the ts parser in their eslint config, right?

Reply:
if they use tseslint in their eslint config, it transforms the typescript AST into an estree AST internally

so the estree `PropertyDefinition` actually just gets some extra properties

we can cast it here because we happen to know the shape, but in future would cast as a `tsestree.PropertyDefinition` or something.

the consumer can use whatever parser they want, it just means when someone uses tseslint (the only way they can be getting `declare` here), we'll understand it

Reply:
ok and then the rule will still be usable for people who are not using the ts parser?

@michellangeveld is this clear enough/actionable for you to continue?

Reply:
Yes added a commit for above